R/kwic.R

```diff
@@ -88,40 +88,47 @@ kwic.corpus <- function(x, pattern, window = 5,
 #'          "Is it a bird or a plane or is it a train?")
 #' kwic(txt, c("is", "a"), valuetype = "fixed")
 #' kwic(txt, phrase(c("is", "a", "is it")), valuetype = "fixed")
-#' 
+#'
 #' toks <- tokens(txt)
 #' kwic(toks, c("is", "a"), valuetype = "fixed")
 #' kwic(toks, phrase(c("is", "a", "is it")), valuetype = "fixed")
-#' 
+#'
 #' corp <- corpus(txt)
 #' kwic(corp, c("is", "a"), valuetype = "fixed", separator = "", remove_separators = FALSE)
-#' @export 
-kwic.tokens <- function(x, pattern, window = 5, 
-                        valuetype = c("glob", "regex", "fixed"), 
-                        separator = " ", 
+#' @export
+kwic.tokens <- function(x, pattern, window = 5,
+                        valuetype = c("glob", "regex", "fixed"),
+                        separator = " ",
                         case_insensitive = TRUE, ...) {
-    
+
     valuetype <- match.arg(valuetype)
     types <- types(x)
+
+    # coerce a character vector into a dictionary
+    if (is.character(pattern)) {
+        patternlist <- as.list(pattern)
+        names(patternlist) <- pattern
+        pattern <- dictionary(patternlist)
+    }
     
     # add document names if none
     if (is.null(names(x))) {
         names(x) <- paste0(quanteda_options("base_docname"), seq_len(x))
     }
-    
-    keywords_id <- pattern2list(pattern, types, 
-                              valuetype, case_insensitive, attr(x, 'concatenator'))
+
+    keywords_id <- pattern2list(pattern, types,
```

Comment:
@koheiw can you assist with this? Here I need to output the `keywords` attribute below as the key from the dictionary that produced the match. This is in the names of `keywords_id` but they are in a different order to the return `temp`. Setting a breakpoint and running this code, I see:
```r
toks <- tokens(c(alpha1 = paste(letters, collapse = " "),
                 alpha2 = paste(LETTERS, collapse = " ")))
quanteda:::kwic.tokens(toks, dictionary(list(ukey = "u", fkey = "f")), window = 3)
Browse[2]> n
debug at /Users/kbenoit/Dropbox (Personal)/GitHub/quanteda/quanteda/R/kwic.R#124: result <- structure(temp, class = c("kwic", "data.frame"), ntoken = ntoken(x), 
    valuetype = valuetype, keywords = names(keywords_id))
Browse[2]> keywords_id
$ukey
[1] 21

$ukey
[1] 47

$fkey
[1] 6

$fkey
[1] 32

attr(,"key")
[1] "ukey" "fkey"
Browse[2]> temp
  docname from to   pre keyword  post
1  alpha1    6  6 c d e       f g h i
2  alpha1   21 21 r s t       u v w x
3  alpha2    6  6 C D E       F G H I
4  alpha2   21 21 R S T       U V W X
```
What I would like is for `keywords` to match the `temp` vector in rows. Probably easiest to return from the C++ function? or to sort beforehand?

Reply:
It is not possible currently to trace the match back to pattern, because we are using `SetNgrams`. 
https://github.com/quanteda/quanteda/blob/612718b1bc2a3d287955850f701e966b002bb145/src/kwic_mt.cpp#L79-L80
But it becomes possible if we change it to `MultiMapNgrams`, which is used in `tokens_lookup()`. However, this means that we have to deal with tricky nested and overlapped matches.

Reply:
Then we need an R function to map the names and numeric type index values in `keywords_id` to `temp$keyword`. But that will be tricky if dictionaries have overlapping values. Or will it - since they will be interchangeable?

Reply:
See `map_keywords()` in `kwic.R`. It works unless pattern is a dictionary and the matches of its values intersect, in which case the `match()` operation will take the first key as its match.

We could warn users about this, since if they want to work around it, then they should perform kwic separately on each (intersecting) key, and then it will work fine for `textplot_xray()`.

Reply:
This is a non-trivial task.

Reply:
Yes I discovered that! We could warn them in the documentation for `textplot_xray()`.